### PR TITLE
8305753: Allow JIT compilation for -Xshare:dump

### PR DESCRIPTION
**Pull request by iklam**
This RFE makes it possible for the user to enable the JIT compiler during the CDS dump by specifying `-Xshare:dump -Xmixed`

By default, `-Xshare:dump` runs with `-Xint`. This maintains the previous behavior.

Which mode to use?

- Dumping in `-Xint` mode is faster when the classlist is small. `-Xint` is also required for deterministic dumping (compiler threads can change the order of the archived object).

- However, dumping in `-Xmixed` mode is faster when the class list is larger. We also need to run the compiler in Leyden so that we can generate the CDS archive and AOT code at the same time.

Tested with tiers 1-4 and builds-tier5.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305753](https://bugs.openjdk.org/browse/JDK-8305753): Allow JIT compilation for -Xshare:dump (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer) ⚠️ Review applies to [0931637e](https://git.openjdk.org/jdk/pull/16233/files/0931637e601f57c17f5793facfca984d2a3018d4)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [0931637e](https://git.openjdk.org/jdk/pull/16233/files/0931637e601f57c17f5793facfca984d2a3018d4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16233/head:pull/16233` \
`$ git checkout pull/16233`

Update a local copy of the PR: \
`$ git checkout pull/16233` \
`$ git pull https://git.openjdk.org/jdk.git pull/16233/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16233`

View PR using the GUI difftool: \
`$ git pr show -t 16233`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16233.diff">https://git.openjdk.org/jdk/pull/16233.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16233#issuecomment-1767690589)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
Is the current use of `-Xint` documented? I'm not seeing how you are communicating this change to the user. To be honest this seems like it is aimed at Leyden, and as such probably belongs over there rather than in mainline.

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -3030,14 +3036,18 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
...
+    if (!mode_flag_cmd_line) {
+      // By default, -Xshare:dump runs in interpreter-only mode, which is required for deterministic archive.
+      //
+      // If your classlist is large and you don't care about deterministic dumping, you can use
+      // -Xshare:dump -Xmixed to improve dumping speed.
+      set_mode_flags(_int);
+    } else if (_mode == _comp) {
+      // -Xcomp may use excessive CPU for the test tiers. Also, -Xshare:dump runs a small and fixed set of
+      // Java code, so there's not much benefit in running -Xcomp.
+      log_info(cds)("reduced -Xcomp to -Xmixed");
```
Comment (matias9927) [inline src/hotspot/share/runtime/arguments.cpp:3048]: Is it preferable to reduce the user-selected option over throwing an error for incompatible options?

File: test/hotspot/jtreg/runtime/cds/DeterministicDump.java [hunk 2]
```diff
@@ -79,6 +79,7 @@ static String dump(ArrayList<String> args, String... more) throws Exception {
         String archiveName = logName + ".jsa";
         String mapName = logName + ".map";
         CDSOptions opts = (new CDSOptions())
+            .addPrefix("-Xint") // Override any -Xmixed/-Xcomp flags from jtreg -vmoptions
```
Comment (matias9927) [inline test/hotspot/jtreg/runtime/cds/DeterministicDump.java:82]: I assume this issue could pop up in newer tests as well. Is there a place where you can specify that `-Xint` needs to be explicitly added when doing a static dump?

File: src/hotspot/share/runtime/arguments.cpp [hunk 3]
```diff
@@ -3030,14 +3036,18 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
...
+      set_mode_flags(_mixed);
```
Comment (matias9927) [inline src/hotspot/share/runtime/arguments.cpp:3049]: Missing an else if case for `_mode==_mixed` ?

**Review by matias9927 — COMMENTED** [review]
Looks good but I have a few questions

**Review by calvinccheung — APPROVED** [review]
Changes seem fine.
Just a suggestion on the log message.
Also, please file a doc subtask if you think it needs to be documented in the release note.

File: src/hotspot/share/runtime/arguments.cpp (hunk 1 above)
Comment (calvinccheung) [inline src/hotspot/share/runtime/arguments.cpp:3048]: I'd suggest the following log message:

`log_info(cds)("reduced -Xcomp to -Xmixed for static dumping");`

Comment (iklam) [reply src/hotspot/share/runtime/arguments.cpp:3048]: We have many test jobs that add `-Xcomp` to every test case, so throwing an error will cause these tests to fail.

If we don't reduce the strength to `-Xmixed`, then many of these tests will fail due to timeout, forcing us to change the timeout value in these tests.

So I think changing to `-Xmixed` is a good compromise to do some testing with the JIT during `-Xshare:dump` without being intrusive.

File: test/hotspot/jtreg/runtime/cds/DeterministicDump.java (hunk 2 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/DeterministicDump.java:82]: -Xint is necessary only when we want to test for deterministic dumps. We don't have many such test cases.

File: src/hotspot/share/runtime/arguments.cpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/runtime/arguments.cpp:3049]: That's intentional: if the user has explicitly chosen -Xint or -Xmixed, we don't override that.

**Comment by iklam** [comment]
> Changes seem fine. Just a suggestion on the log message. Also, please file a doc subtask if you think it needs to be documented in the release note.

I updated the log message as you suggested. I also added the label release-note=yes to the JBS issue.

**Comment by dholmes-ora** [comment]
For this to be useful in mainline, per the example given, then users have to be very aware of its existence. A Release Note is not sufficient, this needs to be documented in the Java command reference (manpage) as well. Thanks.

File: src/hotspot/share/runtime/arguments.cpp (hunk 1 above)
Comment (matias9927) [reply src/hotspot/share/runtime/arguments.cpp]: That makes perfect sense! I am still concerned about the idea taking a user option and reducing it to an option they didn't ask for... I will trust your judgement

**Review by matias9927 — APPROVED** [review]
Thanks for the clarification, this looks good!

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
> For this to be useful in mainline, per the example given, then users have to be very aware of its existence. A Release Note is not sufficient, this needs to be documented in the Java command reference (manpage) as well. Thanks.

Hi David, I added documentation into the java.1 man page.

**Review by dholmes-ora — APPROVED** [review]
Manpage update looks good. Thanks

**Comment by iklam** [comment]
Thanks @matias9927 @calvinccheung @dholmes-ora for the review
/integrate

**Pull request closed without merging**